web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts

```diff
@@ -345,10 +345,22 @@ export class DocumentsService {
     });
   }
 
-  update(uri: string, partialDoc: Partial<Document>) {
+  /**
+   * Updates the document by URI.
+   * @param uri - document URI.
+   * @param updated - a new document object or an update function.
+   */
+  update(
+    uri: DocumentUri,
+    updated: Partial<Document> | ((document: Document) => Document)
```

Comment:
This uses `ImmutableStore` underneath, doesn't it? Shouldn't we accept the function as `(document: Document) => void` then? The current signature makes it seem like it needs to be pure but this doesn't seem to be true.

We would be left with something like this (I haven't checked if this actually works):

```diff
diff --git a/web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts b/web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts
index bf504a4abc..a37086d8cf 100644
--- a/web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts
+++ b/web/packages/teleterm/src/ui/services/workspacesService/documentsService/documentsService.ts
@@ -352,15 +352,15 @@ export class DocumentsService {
    */
   update(
     uri: DocumentUri,
-    updated: Partial<Document> | ((document: Document) => Document)
+    updated: Partial<Document> | ((draftDocument: Document) => void)
   ) {
     this.setState(draft => {
-      const index = draft.documents.findIndex(doc => doc.uri === uri);
-      const toUpdate = draft.documents[index];
+      const toUpdate = draft.documents.find(doc => doc.uri === uri);
       if (typeof updated === 'function') {
-        draft.documents[index] = updated(toUpdate);
+        updated(toUpdate);
+      } else {
+        Object.assign(toUpdate, updated);
       }
-      Object.assign(toUpdate, updated);
     });
   }
 
```

Reply:
Yes, it uses `ImmutableStore`. 
>Shouldn't we accept the function as (document: Document) => void then?

Actually, it should be `(document: Document) => void | Document` (you can return a new state in immer too). I adjusted the implementation for it.